File: packages/client-typescript/src/client/log-codec.ts
/**
 * MIT License
 *
 * Copyright (c) 2026 Aparavi Software AG
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
 * Run-log segment codec — INTERNAL TypeScript mirror of the Python reference
 * (`rocketride._log_codec`). The storage encoding is plumbing, never public
 * SDK surface: this module is deliberately NOT exported from the package
 * barrel — consumers only ever see reconstructed events via the session.
 *
 * Segments are self-contained containers: each opens with a
 * `{"type":"keyframe"}` preamble line, and interior events may carry DELTA
 * bodies whose base is guaranteed to live in the SAME segment (status deltas
 * against the previous status / keyframe base; trace LEAVE deltas against
 * their paired ENTER).
 */

import type { LogEvent } from './types/log.js';

// =============================================================================
// CONSTANTS / TYPES
// =============================================================================

/** Delta body marker key. */
export const DELTA_KEY = '__delta__';

/** Deleted-keys list key inside a delta. */
export const DELETED_KEY = '__deleted__';

/** One open-frame entry inside a keyframe (compact summary + pointers). */
export interface KeyframeOpenFrame {
	id: number | string;
	component?: string;
	doc?: string;
	enterTime?: number;
	enterSeq?: number;
	/** Segment ids that contain this trace's events (sparse expand list). */
	touched?: number[];
}

/** One recently-closed trace summary inside a keyframe. */
export interface KeyframeClosedTrace {
	doc?: string;
	id: number | string;
	/** Begin-event continuum seq — the trace's permanent identity. */
	beginSeq?: number;
	beginTime?: number;
	elapsed?: number;
	calls?: number;
	touched?: number[];
}

/** The segment keyframe preamble (container metadata, not an event). */
export interface SegmentKeyframe {
	type: 'keyframe';
	ver: number;
	/** False on the first keyframe after a writer restart (state unknown). */
	complete: boolean;
	/** True when the open-frame list hit its sanity ceiling. */
	partial: boolean;
	chapter?: { beginSeq?: number; beginTime?: number } | null;
	/** Full status snapshot at the boundary — the status delta base. */
	status: Record<string, unknown>;
	byPipe: Record<string, string[]>;
	openFrames: KeyframeOpenFrame[];
	closedRecent: KeyframeClosedTrace[];
	console: { lines: string[]; truncated: boolean };
}

// =============================================================================
// DELTA APPLICATION
// =============================================================================

/**
 * Inverse of the writer's shallow delta — reconstruct the full object.
 *
 * Dict-valued fields merge one level deep; keys under `__deleted__` are
 * removed. The base is not mutated.
 *
 * @param base - The base object the delta was computed against.
 * @param changes - The changes object from the delta body.
 * @returns The reconstructed full object.
 */
export function applyShallowDelta(base: unknown, changes: unknown): unknown {
	if (typeof base !== 'object' || base === null || typeof changes !== 'object' || changes === null) {
		return changes;
	}
	const result: Record<string, unknown> = { ...(base as Record<string, unknown>) };
	const changeMap = changes as Record<string, unknown>;
	for (const key of (changeMap[DELETED_KEY] as string[] | undefined) ?? []) {
		delete result[key];
	}
	for (const [key, value] of Object.entries(changeMap)) {
		if (key === DELETED_KEY) continue;
		const old = result[key];
		if (value !== null && typeof value === 'object' && !Array.isArray(value) && old !== null && typeof old === 'object' && !Array.isArray(old)) {
			const merged: Record<string, unknown> = { ...(old as Record<string, unknown>) };
			const sub = value as Record<string, unknown>;
			for (const subKey of (sub[DELETED_KEY] as string[] | undefined) ?? []) {
				delete merged[subKey];
			}
			for (const [subKey, subValue] of Object.entries(sub)) {
				if (subKey !== DELETED_KEY) merged[subKey] = subValue;
			}
			result[key] = merged;
		} else {
			result[key] = value;
		}
	}
	return result;
}

// =============================================================================
// SEGMENT DECODER
// =============================================================================

/**
 * Stateful per-segment decoder: resolves delta bodies back to full events.
 *
 * Feed it every line of ONE segment in order (keyframe first). Matching
 * identity for leave deltas mirrors the writer: most-recent open frame of
 * the same (pipe id, component).
 */
export class SegmentDecoder {
	private prevStatus: Record<string, unknown> | null = null;
	/** Per pipe id: stack of [component, enter trace.data as written]. */
	private open = new Map<unknown, Array<[unknown, unknown]>>();

	/** Seed decoder state from a segment's keyframe preamble. */
	seed(keyframe: SegmentKeyframe): void {
		const status = keyframe.status;
		this.prevStatus = status && typeof status === 'object' && Object.keys(status).length > 0 ? status : null;
	}

	/**
	 * Resolve one event's delta body (if any) and update decoder state.
	 *
	 * @param msg - A parsed event line (NOT a keyframe line).
	 * @returns The event with a fully reconstructed body.
	 */
	decode(msg: LogEvent): LogEvent {
		const event = msg.event;
		const body = msg.body as Record<string, unknown> | undefined;

		if (event === 'apaevt_status_update' && body && typeof body === 'object') {
			if (DELTA_KEY in body) {
				const full = applyShallowDelta(this.prevStatus ?? {}, body[DELTA_KEY]) as Record<string, unknown>;
				this.prevStatus = full;
				return { ...msg, body: full as LogEvent['body'] };
			}
			this.prevStatus = body;
			return msg;
		}

		if (event === 'apaevt_flow' && body && typeof body === 'object') {
			const op = body.op;
			const pid = body.id;
			const component = body.component;
			const trace = (body.trace ?? {}) as Record<string, unknown>;
			const data = trace.data;

			if (op === 'begin') {
				if (!this.open.has(pid)) this.open.set(pid, []);
				return msg;
			}
			if (op === 'end') {
				this.open.delete(pid);
				return msg;
			}
			if (op === 'enter') {
				if (!this.open.has(pid)) this.open.set(pid, []);
				this.open.get(pid)!.push([component, data]);
				return msg;
			}
			if (op === 'leave') {
				const stack = this.open.get(pid) ?? [];
				let base: unknown = null;
				for (let i = stack.length - 1; i >= 0; i--) {
					if (stack[i][0] === component) {
						base = stack.splice(i, 1)[0][1];
						break;
					}
				}
				if (data !== null && typeof data === 'object' && DELTA_KEY in (data as Record<string, unknown>)) {
					const fullData = applyShallowDelta(
						base !== null && typeof base === 'object' ? base : {},
						(data as Record<string, unknown>)[DELTA_KEY],
					);
					return { ...msg, body: { ...body, trace: { ...trace, data: fullData } } as unknown as LogEvent['body'] };
				}
				return msg;
			}
		}

		return msg;
	}
}

/**
 * Canonicalize the continuum stamps INTO the body — the single place they
 * live. Current recordings (and the live wire) already carry
 * `body.eventTime` + `body.logSeq`; legacy v2 segments carried the stamps
 * at the header with the continuum under `seq`, so decode moves those into
 * the body once and old data reads identically to new. The DAP envelope is
 * never a source of truth (its `seq` is per-connection bookkeeping).
 *
 * @param msg - A decoded event (mutated in place).
 * @returns The same event with body.eventTime/body.logSeq guaranteed.
 */
export function normalizeStamps(msg: LogEvent): LogEvent {
	let body = msg.body as Record<string, unknown> | undefined;
	if (!body || typeof body !== 'object') {
		body = {};
		(msg as Record<string, unknown>).body = body;
	}
	if (typeof body.eventTime !== 'number' && typeof (msg as Record<string, unknown>).eventTime === 'number') {
		body.eventTime = (msg as Record<string, unknown>).eventTime;
	}
	if (typeof body.logSeq !== 'number' && typeof (msg as Record<string, unknown>).seq === 'number') {
		body.logSeq = (msg as Record<string, unknown>).seq;
	}
	return msg;
}

/**
 * Parse one whole-line-aligned raw segment chunk into keyframe + events,
 * decoding deltas through the supplied decoder.
 *
 * @param text - Raw JSONL text (every line complete — the fetch guarantees it).
 * @param decoder - The segment's decoder (carries state across chunks).
 * @returns The keyframe (when this chunk contained it) and decoded events.
 */
export function parseSegmentChunk(
	text: string,
	decoder: SegmentDecoder,
): { keyframe: SegmentKeyframe | null; events: LogEvent[] } {
	let keyframe: SegmentKeyframe | null = null;
	const events: LogEvent[] = [];
	for (const line of text.split('\n')) {
		if (!line.trim()) continue;
		let msg: Record<string, unknown>;
		try {
			msg = JSON.parse(line) as Record<string, unknown>;
		} catch {
			continue;
		}
		if (msg.type === 'keyframe') {
			keyframe = msg as unknown as SegmentKeyframe;
			decoder.seed(keyframe);
			continue;
		}
		events.push(normalizeStamps(decoder.decode(msg as unknown as LogEvent)));
	}
	return { keyframe, events };
}

File: packages/client-python/src/rocketride/_log_codec.py
# MIT License
#
# Copyright (c) 2026 Aparavi Software AG
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""
Run-log segment codec — the Python reference implementation (DVR v2).

Segments are self-contained containers (video-codec model): each opens with a
``{"type": "keyframe"}`` preamble line carrying accumulated state, and interior
events may carry DELTA bodies whose base is guaranteed to live in the SAME
segment:

- status deltas reference the previous status (or the keyframe's base);
- trace LEAVE deltas reference their paired ENTER (most-recent open frame of
  the same pipe id + component) — leaves whose enter landed in an earlier
  segment are stored full.

This ONE module serves three consumers: the server's RunLogWriter/Reader
(``ai`` depends on this package) and the Python SDK's event-stream session.
The TypeScript SDK mirrors it 1:1.
"""

from typing import Any, Dict, List, Optional

# Delta body marker: {'__delta__': {<changed keys>}}; dict values diff one
# level deep; removed keys are listed under '__deleted__'.
DELTA_KEY = '__delta__'
DELETED_KEY = '__deleted__'


def shallow_delta(prev: Any, curr: Any) -> Any:
    """
    One-level-deep diff of two dicts: changed keys only.

    Dict-valued fields are diffed one level (their changed sub-keys only);
    everything else is compared whole. Keys present in ``prev`` but absent in
    ``curr`` are recorded under ``DELETED_KEY``.

    Args:
        prev: The base object (as previously written to the segment).
        curr: The new object.

    Returns:
        The changes dict (possibly empty), or ``curr`` unchanged when either
        side is not a dict.
    """
    if not isinstance(prev, dict) or not isinstance(curr, dict):
        return curr
    changes: Dict[str, Any] = {}
    for key, value in curr.items():
        old = prev.get(key, DELETED_KEY)
        if isinstance(value, dict) and isinstance(old, dict):
            sub = {k: v for k, v in value.items() if old.get(k, DELETED_KEY) != v}
            sub_deleted = [k for k in old if k not in value]
            if sub or sub_deleted:
                if sub_deleted:
                    sub[DELETED_KEY] = sub_deleted
                changes[key] = sub
        elif old != value:
            changes[key] = value
    deleted = [key for key in prev if key not in curr]
    if deleted:
        changes[DELETED_KEY] = deleted
    return changes


def apply_shallow_delta(base: Any, changes: Any) -> Any:
    """
    Inverse of :func:`shallow_delta` — reconstruct the full object.

    Args:
        base: The base object the delta was computed against.
        changes: The changes dict from the delta body.

    Returns:
        The reconstructed full object (base is not mutated).
    """
    if not isinstance(base, dict) or not isinstance(changes, dict):
        return changes
    result = dict(base)
    for key in changes.get(DELETED_KEY, []):
        result.pop(key, None)
    for key, value in changes.items():
        if key == DELETED_KEY:
            continue
        old = result.get(key)
        if isinstance(value, dict) and isinstance(old, dict):
            merged = dict(old)
            for sub_key in value.get(DELETED_KEY, []):
                merged.pop(sub_key, None)
            for sub_key, sub_value in value.items():
                if sub_key != DELETED_KEY:
                    merged[sub_key] = sub_value
            result[key] = merged
        else:
            result[key] = value
    return result


class SegmentDecoder:
    """
    Stateful per-segment decoder: resolves delta bodies back to full events.

    Feed it every line of ONE segment in order (keyframe first). Matching
    identity for leave deltas mirrors the writer: most-recent open frame of
    the same (pipe id, component).
    """

    def __init__(self) -> None:
        """Start with no base state (a keyframe or first full status seeds it)."""
        self._prev_status: Optional[Dict[str, Any]] = None
        # Per pipe id: stack of (component, enter trace.data as written).
        self._open: Dict[Any, List[tuple]] = {}

    def seed(self, keyframe: Dict[str, Any]) -> None:
        """Seed decoder state from a segment's keyframe preamble."""
        status = keyframe.get('status')
        self._prev_status = status if isinstance(status, dict) and status else None

    def decode(self, msg: Dict[str, Any]) -> Dict[str, Any]:
        """
        Resolve one event's delta body (if any) and update decoder state.

        Args:
            msg: A parsed event line (NOT a keyframe line).

        Returns:
            The event with a fully reconstructed body.
        """
        event = msg.get('event')
        body = msg.get('body')

        if event == 'apaevt_status_update' and isinstance(body, dict):
            if DELTA_KEY in body:
                full = apply_shallow_delta(self._prev_status or {}, body[DELTA_KEY])
                msg = dict(msg)
                msg['body'] = full
                self._prev_status = full
            else:
                self._prev_status = body
            return msg

        if event == 'apaevt_flow' and isinstance(body, dict):
            op = body.get('op')
            pid = body.get('id')
            component = body.get('component')
            trace = body.get('trace') or {}
            data = trace.get('data')

            if op == 'begin':
                self._open.setdefault(pid, [])
                return msg
            if op == 'end':
                self._open.pop(pid, None)
                return msg
            if op == 'enter':
                self._open.setdefault(pid, []).append((component, data))
                return msg
            if op == 'leave':
                stack = self._open.get(pid) or []
                match_idx = next((i for i in range(len(stack) - 1, -1, -1) if stack[i][0] == component), None)
                base = stack.pop(match_idx)[1] if match_idx is not None else None
                if isinstance(data, dict) and DELTA_KEY in data:
                    full_data = apply_shallow_delta(base if isinstance(base, dict) else {}, data[DELTA_KEY])
                    msg = dict(msg)
                    new_trace = dict(trace)
                    new_trace['data'] = full_data
                    new_body = dict(body)
                    new_body['trace'] = new_trace
                    msg['body'] = new_body
                return msg

        return msg


def normalize_stamps(msg):
    """
    Canonicalize the continuum stamps INTO the body — the single place they
    live. Current recordings (and the live wire) already carry
    ``body.eventTime`` + ``body.logSeq``; legacy v2 segments carried the
    stamps at the header with the continuum under ``seq``, so decode moves
    those into the body once and old data reads identically to new. The DAP
    envelope is never a source of truth (its ``seq`` is per-connection
    bookkeeping).

    Args:
        msg: A decoded event dict (mutated in place).

    Returns:
        The same event with body.eventTime/body.logSeq guaranteed.
    """
    body = msg.get('body')
    if not isinstance(body, dict):
        body = {}
        msg['body'] = body
    if not isinstance(body.get('eventTime'), (int, float)) and isinstance(msg.get('eventTime'), (int, float)):
        body['eventTime'] = msg['eventTime']
    if not isinstance(body.get('logSeq'), int) and isinstance(msg.get('seq'), int):
        body['logSeq'] = msg['seq']
    return msg
